packages/compiler/react/evaluator.ts

```diff
@@ -0,0 +1,184 @@
+  const script = new vm.Script(code);
+  const context = vm.createContext(staticContext);
+  const result = script.runInContext(context);
+
+  return result;
+};
+
+export const evaluateUnsafe = (
+  ast: t.Node,
+  scope: NodePath<t.Node>['scope'],
+  excludeIds: string[] = [],
+) => {
+  const bindings: Record<string, any> = scope.getAllBindings() as any;
+  const staticContext: Record<string, any> = {};
+  for (const key in bindings) {
+    try {
+      if (excludeIds.includes(key)) continue;
+      const node = bindings[key].path.node;
+      if (bindings[key].kind === 'const' || bindings[key].kind === 'let') {
+        staticContext[key] = evaluateAstNode(node.init, staticContext);
+      }
+      if (t.isFunctionDeclaration(node) || t.isClassDeclaration(node)) {
+        staticContext[key] = evaluateAstNode(node.body, staticContext);
+      }
+    } catch (_err) {
+      /**/
+    }
+  }
+
+  return getVmValue(ast, staticContext);
+};
+
+export const evaluateAstNode = (
+  ast: t.Node | undefined | null,
+  staticContext: Record<string, any> = {},
+): any => {
+  if (!ast) return undefined;
+
+  if (t.isJSXExpressionContainer(ast)) {
+    return evaluateAstNode(ast.expression, staticContext);
+  }
+
+  if (t.isObjectExpression(ast)) {
+    const ret: Record<string, any> = {};
+    for (let i = -1, len = ast.properties.length; ++i < len; ) {
+      const value = ast.properties[i];
+
+      if (!t.isObjectProperty(value)) {
+        throw new Error('evaluateAstNode can only evaluate object properties');
+      }
+
+      let key: string | number | null | undefined | boolean;
+      if (value.computed) {
+        key = evaluateAstNode(value.key, staticContext);
+      } else if (t.isIdentifier(value.key)) {
+        key = value.key.name;
+      } else if (
+        t.isStringLiteral(value.key) ||
+        t.isNumericLiteral(value.key)
+      ) {
+        key = value.key.value;
+      }
+
+      if (!key || key === true) continue;
+
+      ret[key] = evaluateAstNode(value.value, staticContext);
+    }
+    return ret;
+  }
+
+  if (t.isArrayExpression(ast)) {
+    return ast.elements.map((x) => {
+      return evaluateAstNode(x as any, staticContext);
+    });
+  }
+
+  if (t.isUnaryExpression(ast) && ast.operator === '-') {
+    const ret = evaluateAstNode(ast.argument, staticContext);
+    if (ret === null || ret === undefined) {
+      return null;
+    }
+    return -ret;
+  }
+
+  if (t.isTemplateLiteral(ast)) {
+    let ret = '';
+    for (let idx = -1, len = ast.quasis.length; ++idx < len; ) {
+      const quasi = ast.quasis[idx]!;
+      const expr = ast.expressions[idx];
+      ret += quasi.value.raw;
+      if (expr) {
+        ret += evaluateAstNode(expr, staticContext);
+      }
+    }
+    return ret;
+  }
+
+  if (t.isNullLiteral(ast)) return null;
+
+  if (
+    t.isNumericLiteral(ast) ||
+    t.isStringLiteral(ast) ||
+    t.isBooleanLiteral(ast)
+  ) {
+    return ast.value;
+  }
+
+  if (t.isBinaryExpression(ast)) {
```

Comment:
I noticed the evaluateAstNode could use some enhancements to handle more types of nodes, especially binary operators (such as `**` and `%`). Mind if I directly push some updates to this PR? @aidenybai

Reply:
@poteboy go ahead!